api/src/DataPersister/Util/DataPersisterObservable.php

```diff
@@ -0,0 +1,151 @@
+    }
+
+    public function onCustomAction(CustomActionListener $customActionListener): self {
+        $this->customActionListeners[] = $customActionListener;
+
+        return $this;
+    }
+
+    public function onPropertyChange(PropertyChangeListener $propertyChangeListener): self {
+        $this->propertyChangedListeners[] = $propertyChangeListener;
+
+        return $this;
+    }
+
+    public function persist($data, array $context = []): object {
+        if ('post' === ($context['collection_operation_name'] ?? null)) {
+            $data = call_user_func($this->onBeforeCreate, $data);
+        } elseif ('patch' === ($context['item_operation_name'] ?? null)) {
+            $data = call_user_func($this->onBeforeUpdate, $data);
+        } else {
+            foreach ($this->customActionListeners as $listener) {
+                if (self::actionNameMatches($listener, $context)) {
+                    $data = call_user_func($listener->getBeforeAction(), $data);
+
+                    break;
+                }
+            }
+        }
+
+        $data = $this->dataPersister->persist($data, $context);
+
+        if ('post' === ($context['collection_operation_name'] ?? null)) {
+            call_user_func($this->onAfterCreate, $data);
+        } elseif ('patch' === ($context['item_operation_name'] ?? null)) {
+            call_user_func($this->onAfterUpdate, $data);
+        } else {
+            foreach ($this->customActionListeners as $listener) {
+                if (self::actionNameMatches($listener, $context)) {
+                    call_user_func($listener->getAfterAction(), $data);
+
+                    break;
+                }
+            }
+        }
+
+        $dataBefore = $this->requestStack->getCurrentRequest()->attributes->get('previous_data');
+        if (null != $dataBefore) {
+            foreach ($this->propertyChangedListeners as $listener) {
+                $propertyBefore = call_user_func($listener->getExtractProperty(), $dataBefore);
+                $propertyNow = call_user_func($listener->getExtractProperty(), $data);
+                if ($propertyBefore !== $propertyNow) {
+                    call_user_func($listener->getAfterAction(), $data);
+                }
+            }
+        }
+
+        return $data;
+    }
+
+    public function remove($data, array $context = []) {
+        $beforeRemoveResult = call_user_func($this->onBeforeRemove, $data);
+        if (null == $beforeRemoveResult) {
+            $this->dataPersister->remove($data, $context);
+        } else {
+            $this->dataPersister->persist($data, $context);
+        }
+        call_user_func($this->onAfterRemove, $data);
+    }
+
+    private static function actionNameMatches(CustomActionListener $listener, mixed $context): bool {
+        return $listener->getActionName() === ($context['item_operation_name'] ?? null)
+                || $listener->getActionName() === ($context['collection_operation_name'] ?? null);
```

Comment:
Typisch für API Platform wäre, wenn der Listener nicht einfach einen Namen herausgibt, sondern selber via eine `$listener->supports($context)` Methode oder so sagen kann, ob er am Event interessiert ist. Die Überprüfung rein basierend auf einem Namen (in `item_operation_name`, `collection_operation_name` sowie `graphql_operation_name`) könnte man dann in einer abstrakten Listener-Superklasse oder einem Trait vorprogrammieren.